kimchi/src/circuits/polynomials/keccak/gadget.rs

```diff
@@ -0,0 +1,118 @@
+//! Keccak gate
+
+use ark_ff::PrimeField;
+
+use crate::{
+    alphas::Alphas,
+    circuits::{
+        argument::Argument,
+        expr::E,
+        gate::{CircuitGate, Connect, GateType},
+        lookup::{
+            self,
+            tables::{GateLookupTable, LookupTable},
+        },
+        wires::Wire,
+    },
+};
+
+use super::circuitgates::Xor16;
+
+pub const GATE_COUNT: usize = 1;
+
+impl<F: PrimeField> CircuitGate<F> {
+    /// Create 64-bit word XOR gadget.
+    /// - Generic gate with public input zero to constrain 64-bit length
+    /// - 4 Xor16 for the inputs and output
+    /// - 1 Zero  for completeness
+    ///     
+    /// Outputs tuple (next_row, circuit_gates) where
+    ///  next_row  - next row after this gate
+    ///  gates     - vector of circuit gates comprising this gate
+    pub fn create_xor64(new_row: usize) -> (usize, Vec<Self>) {
+        let gates = vec![
+            CircuitGate {
+                typ: GateType::Xor16,
+                wires: Wire::new(new_row),
+                coeffs: vec![],
+            },
+            CircuitGate {
+                typ: GateType::Xor16,
+                wires: Wire::new(new_row + 1),
+                coeffs: vec![],
+            },
+            CircuitGate {
+                typ: GateType::Xor16,
+                wires: Wire::new(new_row + 2),
+                coeffs: vec![],
+            },
+            CircuitGate {
+                typ: GateType::Xor16,
+                wires: Wire::new(new_row + 3),
+                coeffs: vec![],
+            },
+            CircuitGate {
+                typ: GateType::Zero,
+                wires: Wire::new(new_row + 4),
```

Comment:
Note that here we could use a generic gate to finish the range check on the output (`o = 0`) and then permute the first and second entries with it to check that they are also 0, entirely within the row.